src/main/java/org/visola/cs/datastructure/CircularQueue.java

```diff
@@ -0,0 +1,87 @@
+package org.visola.cs.datastructure;
+
+public class CircularQueue<T> implements Queue<T> {
```

Comment:
A potential simpler implementation could be that you keep track of the queue's size using a separate private instance variable.

The benefit of this is that you don't have to remember to reset the `front` and `rear` markers. All the work the markers have to do is to increment or "wrap".

Whether they should or whether to return the `front` or `rear` value can be determined by the `this.size`.

Reply:
I agree, but the problem was to implement a queue using this "circular algorithm". I'll add other implementations as I progress.
Baby steps.